pkg/server/helpers.go

```diff
@@ -350,3 +353,38 @@ func getUserFromImage(user string) (*int64, string) {
 	// If user is a numeric uid.
 	return &uid, ""
 }
+
+// ensureImageExists returns corresponding metadata of the image reference, if image is not
+// pulled yet, the function will pull the image when ensure is true, and return error when
+// ensure if false.
+func (c *criContainerdService) ensureImageExists(ctx context.Context, ref string, ensure bool) (*metadata.ImageMetadata, error) {
```

Comment:
Can we break it up to two functions instead of relying on `ensure` to toggle the behavior? I think it's generally preferred making the functions more cohesive and easier to reuse.

You could also factor the logic the logic from 361 to 370 to check whether the image exists or not.